src/ctap/mod.rs

```diff
@@ -486,94 +511,126 @@ impl CtapState {
                     }
-                    Command::AuthenticatorReset => self.process_reset(env, channel, now),
-                    Command::AuthenticatorCredentialManagement(params) => {
-                        process_credential_management(
-                            env,
-                            &mut self.stateful_command_permission,
-                            &mut self.client_pin,
-                            params,
-                            now,
-                        )
-                    }
-                    Command::AuthenticatorSelection => self.process_selection(env, channel),
-                    Command::AuthenticatorLargeBlobs(params) => {
-                        self.large_blobs
-                            .process_command(env, &mut self.client_pin, params)
-                    }
-                    Command::AuthenticatorConfig(params) => {
-                        process_config(env, &mut self.client_pin, params)
-                    }
-                    // Vendor specific commands
-                    Command::AuthenticatorVendorConfigure(params) => {
-                        self.process_vendor_configure(env, params, channel)
-                    }
-                    Command::AuthenticatorVendorUpgrade(params) => {
-                        self.process_vendor_upgrade(env, params)
-                    }
-                    Command::AuthenticatorVendorUpgradeInfo => {
-                        self.process_vendor_upgrade_info(env)
-                    }
-                };
-                debug_ctap!(env, "Sending response: {:#?}", response);
-                match response {
-                    Ok(response_data) => {
-                        let mut response_vec = vec![0x00];
-                        if let Some(value) = response_data.into() {
-                            if cbor_write(value, &mut response_vec).is_err() {
-                                response_vec =
-                                    vec![Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR as u8];
-                            }
-                        }
-                        response_vec
-                    }
-                    Err(error_code) => vec![error_code as u8],
                 }
+                response_vec
             }
             Err(error_code) => vec![error_code as u8],
         }
     }
 
+    /// Processed a command after parsing from CBOR, returning its structured output.
+    ///
+    /// This function contains the logic of `parse_command`, minus all CBOR encoding and decoding.
+    /// It should make command parsing easier to test.
+    fn process_parsed_command(
+        &mut self,
+        env: &mut impl Env,
+        command: Command,
+        channel: Channel,
+        now: CtapInstant,
+    ) -> Result<ResponseData, Ctap2StatusCode> {
+        // Correct behavior between CTAP1 and CTAP2 isn't defined yet. Just a guess.
+        #[cfg(feature = "with_ctap1")]
+        {
```

Comment:
I don't think those braces are needed. They're probably a left-over.

Reply:
That would need a `#![feature(stmt_expr_attributes)]`:

> error[E0658]: attributes on expressions are experimental

Reply:
Interesting, this is probably a quirk of our setup. It should work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=19206afa3cb252e5746d4efd0670435c. Let's wait until we have CTAP as a library to figure out where it comes from.

Reply:
This would be the equivalent and it indeed fails to compile: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7f0c21feacafe80a8e5066a838fa701e

Reply:
Oh I see, I didn't know assignment was an expression... that's a bit weird decision, in particular given it returns unit so the expression value isn't useful, only its side-effect. So the work-around would be: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aae7cb8e93d80cfd8f965e808a4677ba. Not sure it's worth it, both solutions are equivalently ugly.

Reply:
This would be much more confusing for me as a reader. I'd feel like I don't understand the purpose. With the current solution, if you don't know the purpose, you're just like: "probably a leftover"